Sources/GRPCCore/Metadata.swift

```diff
@@ -14,5 +14,267 @@
+
+extension Metadata {
+
+  /// An iterator for all string values associated with a given key.
+  /// This iterator will only return values originally stored as strings for a given key.
+  public struct StringValuesIterator: IteratorProtocol {
+    private var metadataIterator: Metadata.Iterator
+    private let key: String
+
+    init(forKey key: String, metadata: Metadata) {
+      self.metadataIterator = metadata.makeIterator()
+      self.key = key
+    }
+
+    public mutating func next() -> String? {
+      while let nextKeyValue = self.metadataIterator.next() {
+        if nextKeyValue.key == self.key {
+          switch nextKeyValue.value {
+          case .string(let stringValue):
+            return stringValue
+          case .binary:
+            continue
+          }
+        }
+      }
+      return nil
+    }
+  }
+
+  /// Create a new ``StringValuesIterator`` that iterates over the string values for the given key.
+  /// - Parameter key: The key over whose string values this iterator will iterate.
+  /// - Returns: An iterator to iterate over string values for the given key.
+  public func makeStringValuesIterator(forKey key: String) -> StringValuesIterator {
+    StringValuesIterator(forKey: key, metadata: self)
+  }
+
+  /// A subscript to get a ``StringValuesIterator`` for a given key.
+  public subscript(keyForStringValues key: String) -> StringValuesIterator {
+    StringValuesIterator(forKey: key, metadata: self)
+  }
+}
+
+extension Metadata {
+
+  /// An iterator for all binary data values associated with a given key.
+  /// This iterator will only return values originally stored as binary data for a given key.
+  public struct BinaryValuesIterator: IteratorProtocol {
+    private var metadataIterator: Metadata.Iterator
+    private let key: String
+
+    init(forKey key: String, metadata: Metadata) {
+      self.metadataIterator = metadata.makeIterator()
+      self.key = key
+    }
+
+    public mutating func next() -> Data? {
+      while let nextKeyValue = self.metadataIterator.next() {
+        if nextKeyValue.key == self.key {
+          switch nextKeyValue.value {
+          case .string:
+            continue
+          case .binary(let binaryValue):
+            return binaryValue
+          }
+        }
+      }
+      return nil
+    }
+  }
+
+  /// Create a new ``BinaryValuesIterator`` that iterates over the `Data` values for the given key.
+  /// - Parameter key: The key over whose `Data` values this iterator will iterate.
+  /// - Returns: An iterator to iterate over `Data` values for the given key.
+  public func makeBinaryValuesIterator(forKey key: String) -> BinaryValuesIterator {
+    BinaryValuesIterator(forKey: key, metadata: self)
+  }
+
+  /// A subscript to get a ``BinaryValuesIterator`` for a given key.
+  public subscript(keyForBinaryValues key: String) -> BinaryValuesIterator {
+    BinaryValuesIterator(forKey: key, metadata: self)
+  }
+}
+
+extension Metadata: ExpressibleByDictionaryLiteral {
```

Comment:
We could actually add `ExpressibleByArrayLiteral` conformance too... sometimes it's useful to have